src/util/shared_exclusive_lock.c

```diff
@@ -12,98 +12,146 @@
+// GILAlternativeLock must be held when waiting on this condition.
+pthread_cond_t GILAvailable = PTHREAD_COND_INITIALIZER;
+// Condition for the main thread to wait on while trying to take back the GIL.
+pthread_cond_t GILIsBorrowed = PTHREAD_COND_INITIALIZER;
+// Flags to indicate whether the GIL is lent by the main thread
+bool GIL_lent = false;
+// Flag to indicate whether the GIL is currently borrowed by any thread
+bool GIL_borrowed = false;
+
+// Locks Order:
+// 1. GIL
+// 2. GILAlternativeLock
+// 3. InternalLock
+// A thread may hold any combination of these locks, but must always acquire them in this order.
+// One exception is that we acquire the GIL last in SharedExclusiveLock_Acquire, but there is
+// no risk of deadlock because we do it with a "try" mechanism.
+
+#define TIMEOUT_NANOSECONDS 5000 // 5 us in nanoseconds
+#define NANOSEC_PER_SECOND 1000000000L // 10^9
 
+static inline void set_timeout(struct timespec *timeout) {
+  clock_gettime(CLOCK_MONOTONIC_RAW, timeout);
+  // Assumes TIMEOUT_NANOSECONDS will not exceed NANOSEC_PER_SECOND,
+  // so we are only off by one second maximum.
+  timeout->tv_nsec += TIMEOUT_NANOSECONDS;
+  if (timeout->tv_nsec >= NANOSEC_PER_SECOND) {
+    timeout->tv_nsec -= NANOSEC_PER_SECOND;
+    timeout->tv_sec += 1;
+  }
+}
 
 void SharedExclusiveLock_Init() {
-  GILOwned = false;
-  GILAlternativeLockHeld = false;
+  GIL_lent = false;
+  GIL_borrowed = false;
   pthread_mutex_init(&GILAlternativeLock, NULL);
   pthread_mutex_init(&InternalLock, NULL);
-  pthread_cond_init(&TryLockCondition, NULL);
-  pthread_cond_init(&GILSafeCondition, NULL);
+
+  // Initialize GILAvailable with CLOCK_MONOTONIC_RAW
+  pthread_condattr_t cond_attr;
+  pthread_condattr_init(&cond_attr);
+  pthread_condattr_setclock(&cond_attr, CLOCK_MONOTONIC_RAW);
+  pthread_cond_init(&GILAvailable, &cond_attr);
+  pthread_condattr_destroy(&cond_attr);
+
+  pthread_cond_init(&GILIsBorrowed, NULL);
 }
 
 void SharedExclusiveLock_Destroy() {
   pthread_mutex_destroy(&GILAlternativeLock);
-  pthread_cond_destroy(&TryLockCondition);
+  pthread_cond_destroy(&GILAvailable);
   pthread_mutex_destroy(&InternalLock);
-  pthread_cond_destroy(&GILSafeCondition);
+  pthread_cond_destroy(&GILIsBorrowed);
 }
 
-void SharedExclusiveLock_SetOwned() {
+// Assumptions:
+// 1. The caller holds the GIL.
+// 2. The caller won't release the GIL before calling TakeBackGIL.
+// Note: The caller may call SharedExclusiveLock_Acquire while holding the GIL, if it lends it.
+void SharedExclusiveLock_LendGIL() {
   pthread_mutex_lock(&InternalLock);
-  GILOwned = true;
-  // Signal any waiting threads that they may try to acquire the GIL or the alternative lock.
-  pthread_cond_broadcast(&TryLockCondition);
+  GIL_lent = true;
   pthread_mutex_unlock(&InternalLock);
+  // Signal any waiting threads that they may try to borrow the GIL.
+  pthread_cond_broadcast(&GILAvailable);
```

Comment:
Are we sure about this pattern? didn't we say to broadcast before unlocking?

Reply:
In this case there is no relation between the lock and condition. We do broadcast the signal before releasing `GILAlternativeLock`